kong/tools/uri.lua

```diff
@@ -44,20 +31,38 @@ local SLASH = string_byte("/")
 local function percent_decode(m)
     local hex = m[1]
     local num = tonumber(hex, 16)
-    if RESERVED_CHARACTERS[num] then
-      return string_upper(m[0])
+    -- from rfc3986: ...should be normalized by decoding any
+    -- percent-encoded octet that corresponds to an unreserved character
+    --
+    -- unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
+    if (num >= A_BYTE and num <= Z_BYTE) -- a..z
+       or (num >= CAP_A_BYTE and num <= CAP_Z_BYTE) -- A..Z
+       or (num >= ZERO_BYTE and num <= NINE_BYTE) -- 0..9
+       or num == HYPHEN_BYTE or num == DOT_BYTE
+       or num == UNDERSCORE_BYTE or num == TILDE_BYTE
+    then
+      return string_char(num)
     end
 
-    return string_char(num)
+    return string_upper(m[0])
 end
 
 
-local function escape(m)
+local function percent_escape(m)
   return string_format("%%%02X", string_byte(m[0]))
 end
 
 
-function _M.normalize(uri, merge_slashes)
+local function escape(uri)
+  if ngx_re_find(uri, ESCAPE_PATTERN, "joi") then
+    return ngx_re_gsub(uri, ESCAPE_PATTERN, percent_escape, "joi")
+  end
+
+  return uri
+end
+
+
+local function normalize(uri, merge_slashes)
```

Comment:
This function does not escape anymore.

Reply:
This issue can be illustrated with curl:
```bash
~ ❯ curl http://localhost:8080/uri/ö      
<html><body>/uri/ö</html>
```

where I have this location block:
```
location /uri {
  content_by_lua '
    ngx.say("<html><body>", ngx.var.request_uri ,"</html>")
  ';
}
```

See, the `ö` is unescaped, but normalization function has to escape it properly.